make torch.testing error messages more expressive

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56373 add `torch.testing` to docs
* #56365 Split `test_testing.py::TestAsserts` for multiple devices
* #55890 enable support for custom error messages in `torch.testing`
* #55788 add support for equal_nan in torch.testing.assert_close
* #55786 Add support for scalars and numpy in torch.testing
* #55385 Add support for checking tensor containers in `torch.testing`
* **#55145 make torch.testing error messages more expressive**

Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

Differential Revision: [D27877157](https://our.internmc.facebook.com/intern/diff/D27877157)

pmeier committed:
make torch.testing error messages more expressive

Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier commented:
From @mruberry in https://github.com/pytorch/pytorch/pull/54784#issuecomment-816385781

> Brief elaboration on this topic:
>
> It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs. Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.

pmeier commented:
> It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs.

Not sure what you mean there. Isn't the index doing exactly that? In the example above `a[2, 3]` and `b[2, 3]` is the combination that results in the maximum absolute and relative difference.

> Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.

I've removed just that, albeit private, in this PR: `_trace_mismatches`. Do you want something like that as public interface? If so, how would it look like? `trace_(equal|close)_mismatches(a, b, **kwargs)`? What would it return? A named tuple with all the information?

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

mruberry commented:
> > It'd be really nice in the pointwise numerics tests if we could relate the indices of greatest difference back to the inputs.
> 
> Not sure what you mean there. Isn't the index doing exactly that? In the example above `a[2, 3]` and `b[2, 3]` is the combination that results in the maximum absolute and relative difference.

Yes but it'd be nice to do that progammatically so the test could acquire the inputs and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".

> > Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.
> 
> I've removed just that, albeit private, in this PR: `_trace_mismatches`. Do you want something like that as public interface? If so, how would it look like? `trace_(equal|close)_mismatches(a, b, **kwargs)`? What would it return? A named tuple with all the information?

I'm not sure yet. Let's focus on getting the core right and then we can consider additions like this.

mruberry commented on test/test_testing.py:
```diff
@@ -798,42 +798,80 @@ def test_mismatching_values(self, device):
     @onlyCPU
     def test_mismatching_values_msg_abs_mismatches(self, device):
-        a = torch.empty((3, 4), dtype=torch.float32, device=device).fill_(5)
+        a = torch.empty((3, 3), dtype=torch.float32, device=device).fill_(5)
```
Prefer zeros to empty, since empty could be anything (including the correct value)

pmeier replied:
1. Shouldn't this make no difference given that we are invoking `fill_()` after the instantiation?
2. Why do you want to change the size?

mruberry replied:
Good point. I missed the fill_, sorry.

In that case uses torch.full().

I'm not sure I understand what you mean about changing the size?

pmeier replied:
> I'm not sure I understand what you mean about changing the size?

My bad, I misread the diff.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
             location of mismatches.
+
+    Returns:
+        (str): Multiline list comprising the absolute and relative number of mismatches, and the maximum absolute and
```
A string that contains the greatest absolute difference between elements of a and b, the greatest relative difference between elements of a and b, and the indices of these differences.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
-    rel = abs / total
+    total_elements = mismatches.numel()
+    abs_mismatches = torch.sum(mismatches).item()
+    rel_mismatches = abs_mismatches / total_elements
```
Because "relative" is so overloaded this should be the percentage of mismatched elements

pmeier replied:
But percentage would also not be right unless we multiply by 100 here, right?

pmeier replied:
I went for `mismatch_ratio` instead.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
-    abs = torch.sum(mismatches).item()
-    rel = abs / total
+    total_elements = mismatches.numel()
+    abs_mismatches = torch.sum(mismatches).item()
```
Because "absolute" is so overloaded this would be the total number of mismatched elements

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+    rel_mismatches = abs_mismatches / total_elements
 
     dtype = torch.float64 if a.dtype.is_floating_point else torch.int64
     a_flat = a.flatten().to(dtype)
```
Why flatten?

pmeier replied:
To be able to easily find the maximum. [`torch.max`](https://pytorch.org/docs/stable/generated/torch.max.html#torch-max) operates on one dimension. If we would pass in an input with more dimensions we would get a dimension-wise maximum here.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+
+    return (
+        f"Mismatched elements: {abs_mismatches} / {total_elements} ({rel_mismatches:.1%})\n"
+        f"Max. abs. diff.: {max_abs_diff.item()} at {_unravel_index(max_abs_diff_flat_idx.item(), mismatches.shape)}\n"
```
"Greatest absolute difference"

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+    return (
+        f"Mismatched elements: {abs_mismatches} / {total_elements} ({rel_mismatches:.1%})\n"
+        f"Max. abs. diff.: {max_abs_diff.item()} at {_unravel_index(max_abs_diff_flat_idx.item(), mismatches.shape)}\n"
+        f"Max. rel. diff.: {max_rel_diff.item()} at {_unravel_index(max_rel_diff_flat_idx.item(), mismatches.shape)}"
```
"Greatest relative difference:"

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+    abs_diff = torch.abs(a_flat - b_flat)
+    max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
+
+    rel_diff = abs_diff / torch.abs(b_flat)
```
This is an interesting way of computing the greatest relative difference, because it ignores the magnitude of elements of a. 

What if a is large but b is small? Won't that report a huge relative error but if a is small and b is large the relative error would be small?

Can we consider being symmetric? I understand that NumPy's actual comparison operations are asymmetric. I wonder if that's changing in the array API?

pmeier replied:
Note, that [`torch.isclose`](https://pytorch.org/docs/stable/generated/torch.isclose.html?highlight=isclose#torch.isclose) and thus in turn `assert_tensors_close` is also asymmetric. This way `rel_diff` is the minimal value for `rtol` that would make the check pass.

> Can we consider being symmetric?

Sure, how would you go about it? Compute the relative difference both ways and pick the smaller one?

> I understand that NumPy's actual comparison operations are asymmetric.

Not sure what you mean by "actual comparison". If you have a look at [`numpy.testing.assert_allclose`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy-testing-assert-allclose) they are using the same logic to find mismatches. They named the parameters `actual` and `desired` to make this more clear. 

I'll check if they do anything different for their error message.

mruberry replied:
You're right. I guess if we keep it this way then the atol and rtol are just these values, because the check is consistent with close.

What about "actual" and "expected" for names? It'd be nice to clarify to users which tensor should go where.

pmeier replied:
Can do that. Do you want this naming scheme only for `assert_close` or everywhere for consistency? `assert_equal` is symmetric.

mruberry replied:
What do you think?

pmeier replied:
I would do it everywhere for consistency. Since the first two inputs should be positional only (although they can't be right now), this should not make it more verbose for the users. Plus, in case we want something asymmetric for `assert_equal` we are on the safe side.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -243,11 +227,8 @@ def _check_values_close(
-        f"With rtol={rtol} and atol={atol}, "
-        f"found {trace.abs} different element(s) out of {trace.total} ({trace.rel:.1%}). "
-        f"The greatest difference of {trace.diff} ({trace.a} vs. {trace.b}) occurred at index {trace.idx}"
+        f"Tensors are not close according to rtol={rtol} and ator={atol}:\n\n{_trace_mismatches(a, b, mismatches)}"
```
"ator" -> "atol"

mruberry replied:
"Tensors are not close:"

or

"Tensors are not close!" 

(see above)

mruberry replied:
You're correct that if we follow the current isclose() logic we don't need to take this recommendation, because the atol and rtol needed are just the greatest absolute difference and greatest relative difference. Maybe instead we can say:

"Greatest absolute difference: {x} (up to {y} allowed)"

or:

"
Greatest absolute difference: {x}
Allowed absolute difference: {y}
"

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,49 +154,37 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+
+    return (
+        f"Mismatched elements: {abs_mismatches} / {total_elements} ({rel_mismatches:.1%})\n"
+        f"Max. abs. diff.: {max_abs_diff.item()} at {_unravel_index(max_abs_diff_flat_idx.item(), mismatches.shape)}\n"
```
It might be interesting to return these statistics as data. Then the close assertion could create its own string that inserted the greatest absolute difference and greatest relative difference that would be allowed.

pmeier replied:
So something like https://github.com/pytorch/pytorch/pull/55145#issuecomment-817297273?

pmeier commented:
> Yes but it'd be nice to do that progammatically so the test could acquire the indices and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".

We could add a `msg` parameter to our public interface. If it is used, every error message will simply replaced by it. 

To give the user more options, we could also accept a callable that will be called with the mismatch statistics and returns the message of the error to be raised.

mruberry commented:
> > Yes but it'd be nice to do that progammatically so the test could acquire the indices and apply them to the input tensor and edit its error message to say: "here is the input where torch.cos() and np.cos() had wildly different values".
> 
> We could add a `msg` parameter to our public interface. If it is used, every error message will simply replaced by it.
> 
> To give the user more options, we could also accept a callable that will be called with the mismatch statistics and returns the message of the error to be raised.

Maybe. Let's talk about it more for a future change.

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,58 +158,71 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
+    - total_elements (int): Total number of values.
+    - total_mismatches (int): Total number of mismatches.
+    - mismatch_ratio (float): Quotient of total mismatches and total elements.
+    - max_abs_diff (Union[int, float]): Greatest absolute difference of :attr:`a` and :attr:`b`.
```
This looks really good; we don't need to block this diff in the stack on this question, but is NaN vs 0 a greater absolute difference than inf vs 0? What about 5 vs 0?

pmeier replied:
```python
>>> torch.sort(torch.tensor([5, float("inf"), float("nan")])).values
tensor([5., inf, nan])
```

This means in case any NaN or Inf value is present it will always be used as greatest diff. We should definitively revisit this in the future.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -154,58 +158,71 @@ def _equalize_attributes(a: torch.Tensor, b: torch.Tensor) -> Tuple[torch.Tensor
 
 
-_Trace = namedtuple("Trace", ("total", "abs", "rel", "idx", "diff", "a", "b"))
+_Trace = namedtuple(
```
This is cool, especially since it will be easy to extend in the future if we like

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+    Args:
+        a (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): First input.
+        b (Union[Tensor, Sequence[Tensor], Mapping[Any, Tensor]]): Second input.
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Callable used to check if two tensors match.
```
This is a nice design

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+    """
+    # _check_are_tensors() returns nothing in case both inputs are tensors and an exception otherwise. Thus, the logic
+    # is inverted here.
+    are_tensors = not _check_are_tensors(a, b)
```
Might as well check explicitly for this being None

Is it weird that it doesn't return a tuple (True/False, optional exception)?

pmeier replied:
Returning `ok:bool, exc: Optional[Exception]` has redundant information since the output would always be one of `True, Exception` or `False, None`. Thus, `ok == not(bool(exc))`. This discussion is moot anyway, because #55786 removes this check in order to allow other input types.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+    return None
+
+
+_MAPPING_MSG_FMTSTR = "The failure occurred for key '{}' of the mappings."
```
It's great to propagate the origins of the failures like this

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
+        additional_keys = a_keys - b_keys
+        return AssertionError(
+            f"The keys of the mappings do not match:\n\n"
+            f"Missing keys in the first mapping: {sorted(missing_keys)}\n"
```
This is cool and we shouldn't block this diff on this question, but this might be more natural with "expected" and "actual" vs. "first"

pmeier replied:
Resolves automatically with https://github.com/pytorch/pytorch/pull/55145#discussion_r613042365

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -245,118 +264,350 @@ def _check_values_close(
-            :attr:`~torch.Tensor.dtype`, they are copied to the :class:`~torch.dtype` returned by
-            :func:`torch.promote_types` before their values are compared.
-        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same stride.
+        check_device (bool): If ``True`` (default), asserts that tensors live in the same :attr:`~torch.Tensor.device`
```
nit: "that both tensors are on the same device" would be more consistent with other PyTorch language

mruberry replied:
(because tensors, like vampires, aren't really alive)

mruberry reviewed (approved):
Awesome!

At this point I think the issues with this diff are so minor or debatable that we should revisit them in future diffs to maintain code velocity.

For example, "actual" and "expected" is a simple refinement to perform later, and I don't want slow down the landing process for it.

mruberry commented:
Would you push a trivial update to this branch, @pmeier, to retrigger the lint builds?

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

Differential Revision: [D27877157](https://our.internmc.facebook.com/intern/diff/D27877157)

[ghstack-poisoned]

pmeier committed:
Update on "make torch.testing error messages more expressive"


Repeating the discussion from https://github.com/pytorch/pytorch/pull/54784#issuecomment-811792089

The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

This implements a mix of both to get a short but expressive message:

```
Tensors are not close according to rtol=1.3e-6 and atol=1e-05:

Mismatched elements: 1 / 12 (8.3%)
Max. rel. diff.: 4.44e-1 at (2, 3)
Max. abs. diff.: 4.0 at (2, 3)
```

Differential Revision: [D27877157](https://our.internmc.facebook.com/intern/diff/D27877157)

[ghstack-poisoned]
